### PR TITLE
fix: capture PerformanceNavigationTiming events even when window.load fires before plugin loads

### PR DESCRIPTION
**Pull request by adebayor123**
Modified NavigationPlugin to capture PerformanceNavigationEvents when the plugin loads after the page does.
- Removed PerformanceObserver from eventListener.
- Added two new unit test cases and modified existing one case with respect to the new implementation change.
- Created a new app called delayed_page.html that loads the script after the page finishes loading.
- Created a new integ test case using the delayed_page app to verify that PerformanceNavigationEvents are captured.

To test the change, do the following:
- Create a temporary app (delayed_page.html) in your local repository without any other implementation changes
- Run server and either use Chrome debugger or open up local host, compare the output between index.html and delayed_page.html.
- index.html should have the navigation events, while the delayed_page.html does not.
- This indicates that with the original implementation, delayed_page properly simulates a web page loading prior to plugin
- Once this is verified, pull in the NavigationPlugin changes, and rerun the server/debugger to see that now PerformanceNavigationEvents are captured in the delayed_page app.
---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/plugins/event-plugins/NavigationPlugin.ts [hunk 1]
```diff
@@ -47,6 +57,24 @@ export class NavigationPlugin implements Plugin {
...
+    /**
+     * Use the loadEventEnd field from window.performance to check if the website
+     * has loaded already.
+     * @returns boolean
+     */
+    hasTheWindowLoadEventFired() {
+        if (
+            window.performance &&
+            window.performance.getEntriesByType(NAVIGATION).length
+        ) {
```
Comment (qhanam) [inline src/plugins/event-plugins/NavigationPlugin.ts:69]: Does this guarantee that `window.load` has fired? 

Or should we look for a specific event that signals the page has finished loading -- i.e., [`PerformanceTiming.loadEventEnd`](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceTiming/loadEventEnd) or [`PerformanceNavigationTiming.loadEventEnd`](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceNavigationTiming/loadEventEnd)

File: src/plugins/event-plugins/__tests__/NavigationPlugin.test.ts [hunk 2]
```diff
@@ -92,13 +93,43 @@ describe('NavigationPlugin tests', () => {
...
 
         // Assert
         expect(record).toHaveBeenCalledTimes(0);
     });
+
+    test('when page has not loaded then we do not fire until load is completed', async () => {
+        // Setting up new mocked window that has not loaded
+        (window as any).performance = performanceEventNotLoaded.performance();
+        (window as any).PerformanceObserver =
+            performanceEvent.PerformanceObserver;
```
Comment (qhanam) [inline src/plugins/event-plugins/__tests__/NavigationPlugin.test.ts:112]: Do we need to mock `PerformanceObserver` here? It was removed from `NavigationPlugin`

File: src/plugins/event-plugins/NavigationPlugin.ts [hunk 3]
```diff
@@ -62,18 +90,9 @@ export class NavigationPlugin implements Plugin {
...
-                            event
-                        );
-                    }
-                });
-            });
-            navigationObserver.observe({
-                entryTypes: [NAVIGATION]
-            });
+            this.performanceNavigationEventHandlerTimingLevel2(
+                performance.getEntriesByType(NAVIGATION)[0]
```
Comment (qhanam) [inline src/plugins/event-plugins/NavigationPlugin.ts:94]: Is it guaranteed that there is only one navigation timing entry?

File: src/plugins/event-plugins/__integ__/NavigationPlugin.test.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+const COMPRESSION_RATIO = 'compressionRatio';
+const FILE_TYPE = 'fileType';
+const SAFARI = 'Safari';
+const FIREFOX = 'Firefox';
+
+fixture('NagivationEvent Plugin').page(
+    'http://localhost:8080/delayed_page.html'
+);
+
+test('Delayed plugin loading still records load events', async (t: TestController) => {
```
Comment (qhanam) [inline src/plugins/event-plugins/__integ__/NavigationPlugin.test.ts:55]: nit: "when plugin loads after window.load then navigation timing events are recorded"

File: src/plugins/event-plugins/__tests__/NavigationPlugin.test.ts [hunk 5]
```diff
@@ -92,13 +93,43 @@ describe('NavigationPlugin tests', () => {
...
+        plugin.load(context);
+        // assert that the plugin did not fire
+        expect(record).toHaveBeenCalledTimes(0);
+
+        // now that the page has loaded, we should fire
+        window.dispatchEvent(new Event('load'));
+        expect(record).toHaveBeenCalledTimes(1);
+    });
+
+    test('when page loads before then we still fire when loading plugin', async () => {
```
Comment (qhanam) [inline src/plugins/event-plugins/__tests__/NavigationPlugin.test.ts:125]: nit: 'when window.load fires before plugin loads then navigation timing is recorded'

File: src/plugins/event-plugins/__tests__/NavigationPlugin.test.ts [hunk 6]
```diff
@@ -92,13 +93,43 @@ describe('NavigationPlugin tests', () => {
...
+        plugin.load(context);
+
         window.dispatchEvent(new Event('load'));
         plugin.disable();
 
         // Assert
         expect(record).toHaveBeenCalledTimes(0);
     });
+
+    test('when page has not loaded then we do not fire until load is completed', async () => {
```
Comment (qhanam) [inline src/plugins/event-plugins/__tests__/NavigationPlugin.test.ts:108]: nit: 'when window.load fires after plugin loads then navigation timing is recorded'

**Comment by qhanam** [comment]
Please modify the commit message to use the following conventions:
1. Use [conventional commit format](https://www.conventionalcommits.org/)
    i.e.,\<type\>[optional scope]: \<description\>
2. Use [the imperative mood in the subject line](https://cbea.ms/git-commit/#imperative)
i.e., a commit message subject line should be able to complete the sentence _"If applied, this commit will __commit message subject line__"_

File: src/plugins/event-plugins/__tests__/NavigationPlugin.test.ts (hunk 2 above)
Comment (adebayor123) [reply src/plugins/event-plugins/__tests__/NavigationPlugin.test.ts:112]: Agreed, no need. Will reflect on next revision.

File: src/plugins/event-plugins/NavigationPlugin.ts (hunk 3 above)
Comment (adebayor123) [reply src/plugins/event-plugins/NavigationPlugin.ts:94]: Good catch, I will change it to iterate instead.

File: src/plugins/event-plugins/NavigationPlugin.ts (hunk 1 above)
Comment (adebayor123) [reply src/plugins/event-plugins/NavigationPlugin.ts:69]: Discussed offline; The following line to the if statement uses PerformanceNavigationTiming.loadEventEnd

File: src/plugins/event-plugins/NavigationPlugin.ts (hunk 3 above)
Comment (adebayor123) [reply src/plugins/event-plugins/NavigationPlugin.ts:94]: Actually, just found this: https://www.w3.org/TR/navigation-timing-2/ 
"Only the [current document](https://www.w3.org/TR/navigation-timing-2/#dfn-current-document) resource is included in the performance timeline; there is only one [PerformanceNavigationTiming](https://www.w3.org/TR/navigation-timing-2/#dom-performancenavigationtiming) object in the performance timeline."

**Review by qhanam — APPROVED** [review]

**Pull request merged**